cpp/src/arrow/compute/exec/union_node_test.cc

```diff
@@ -0,0 +1,203 @@
+  auto l_schema = schema({field("colum_i32", int32()), field("colum_str", utf8())});
+  auto r_schema = schema({field("colum_i32", int32()), field("colum_str", utf8())});
+  BatchesWithSchema l_batches, r_batches, exp_batches;
+
+  int multiplicity = parallel ? 100 : 1;
+
+  GenerateBatchesFromString(l_schema,
+                            {
+                                R"([[0,"d"], [1,"b"]])",
+                                R"([[2,"d"], [3,"a"], [4,"a"]])",
+                            },
+                            &l_batches, multiplicity);
+
+  GenerateBatchesFromString(r_schema,
+                            {
+                                R"([[10,"A"]])",
+                            },
+                            &r_batches, multiplicity);
+
+  GenerateBatchesFromString(l_schema,
+                            {
+                                R"([[0,"d"], [1,"b"]])",
+                                R"([[2,"d"], [3,"a"], [4,"a"]])",
+
+                                R"([[10,"A"]])",
+                            },
+                            &exp_batches, multiplicity);
+  CheckRunOutput(l_batches, r_batches, exp_batches, parallel);
+}
+
+void RunEmptyTest(bool parallel) {
+  auto l_schema = schema({field("colum_i32", int32()), field("colum_str", utf8())});
+  auto r_schema = schema({field("colum_i32", int32()), field("colum_str", utf8())});
+
+  int multiplicity = parallel ? 100 : 1;
+
+  BatchesWithSchema l_empty, r_empty, output_batches;
+
+  GenerateBatchesFromString(l_schema, {R"([])"}, &l_empty, multiplicity);
+  GenerateBatchesFromString(r_schema, {R"([])"}, &r_empty, multiplicity);
+
+  GenerateBatchesFromString(l_schema, {R"([])", R"([])"}, &output_batches, multiplicity);
+
+  CheckRunOutput(l_empty, r_empty, output_batches);
+}
+
+TEST(UnionTest, TestNonEmpty) {
+  for (bool parallel : {false, true}) {
+    RunNonEmptyTest(parallel);
+  }
+}
+
+TEST(UnionTest, TestEmpty) {
+  for (bool parallel : {false, true}) {
+    RunEmptyTest(parallel);
+  }
+}
+
+void TestUnionRandom(const std::shared_ptr<DataType>& data_type, bool parallel,
+                     int num_batches, int batch_size) {
+  auto l_schema = schema({field("colum0", data_type), field("colum1", data_type)});
+  auto r_schema = schema({field("colum0", data_type), field("colum1", data_type)});
+
+  // generate data
+  auto l_batches = MakeRandomBatches(l_schema, num_batches, batch_size);
+  auto r_batches = MakeRandomBatches(r_schema, num_batches, batch_size);
+
+  ASSERT_OK_AND_ASSIGN(auto plan, ExecPlan::Make());
+
+  Declaration Union{"union", ExecNodeOptions{}};
+
+  // add left source
+  Union.inputs.emplace_back(Declaration{
+      "source", SourceNodeOptions{l_batches.schema, l_batches.gen(parallel,
+                                                                  /*slow=*/false)}});
+  // add right source
+  Union.inputs.emplace_back(Declaration{
+      "source", SourceNodeOptions{r_batches.schema, r_batches.gen(parallel,
```

Comment:
Maybe we could make this test use three inputs? We could even just duplicate one of the existing inputs.